File: leetcode/valid-anagram.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        if (s.length() != t.length()) return false;

        map<char, int> letter_freq;
        for (char c: s)
        {
            letter_freq[c]++;
        }

        for (char c: t)
        {
            if (letter_freq.find(c) == letter_freq.end()) return false;
            else if (letter_freq[c] < 1) return false;
            else letter_freq[c]--;
        }
        return true;
    }
};

File: leetcode/valid-anagram.py
class Solution:
    ## O(n logn)
    # def isAnagram(self, s: str, t: str) -> bool:
    #     return sorted(s) == sorted(t)

    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t): return False

        freqs = {}
        for letter in s:
            if letter not in freqs:
                freqs[letter] = 0
            freqs[letter] += 1
        
        for letter in t:
            if letter not in freqs: return False
            elif freqs[letter] < 1: return False
            else: freqs[letter] -= 1
        
        return True
